clang/lib/CConv/RewriteUtils.cpp

```diff
@@ -953,128 +586,36 @@ std::string ArrayBoundsRewriter::getBoundsString(PVConstraint *PV,
 void RewriteConsumer::HandleTranslationUnit(ASTContext &Context) {
   Info.enterCompilationUnit(Context);
 
-  // Compute the bounds information for all the array variables.
-  ArrayBoundsRewriter ABRewriter(&Context, Info);
-
+  // Rewrite Variable declarations
   Rewriter R(Context.getSourceManager(), Context.getLangOpts());
-  std::set<FileID> Files;
-
-  std::set<std::string> v;
-  RSet RewriteThese(DComp(Context.getSourceManager()));
-  // Unification is done, so visit and see if we need to place any casts
-  // in the program.
-  TypeRewritingVisitor TRV = TypeRewritingVisitor(&Context, Info, RewriteThese, v,
-                                                  RewriteConsumer::
-                                                      ModifiedFuncSignatures,
-                                                  ABRewriter);
-  for (const auto &D : Context.getTranslationUnitDecl()->decls())
-    TRV.TraverseDecl(D);
-
-  // Build a map of all of the PersistentSourceLoc's back to some kind of
-  // Stmt, Decl, or Type.
-  VariableMap &VarMap = Info.getVarMap();
-  std::set<PersistentSourceLoc> keys;
-
-  for (const auto &I : VarMap)
-    keys.insert(I.first);
-  SourceToDeclMapType PSLMap;
-  VariableDecltoStmtMap VDLToStmtMap;
-
-  RSet skip(DComp(Context.getSourceManager()));
-  MappingVisitor V(keys, Context);
-  TranslationUnitDecl *TUD = Context.getTranslationUnitDecl();
-  StructVariableInitializer FV =
-      StructVariableInitializer(&Context, Info, RewriteThese);
-  GlobalVariableGroups GVG(R.getSourceMgr());
-  std::set<llvm::FoldingSetNodeID> seen;
-  std::map<llvm::FoldingSetNodeID, AnnotationNeeded> nodeMap;
-  CheckedRegionFinder CRF(&Context, R, Info, seen, nodeMap);
-  CheckedRegionAdder CRA(&Context, R, nodeMap);
+  std::set<FileID> TouchedFiles;
+  DeclRewriter::rewriteDecls(Context, Info, R, TouchedFiles);
+
+  // Take care of some other rewriting tasks
+  std::set<llvm::FoldingSetNodeID> Seen;
+  std::map<llvm::FoldingSetNodeID, AnnotationNeeded> NodeMap;
+  CheckedRegionFinder CRF(&Context, R, Info, Seen, NodeMap);
+  CheckedRegionAdder CRA(&Context, R, NodeMap);
   CastPlacementVisitor ECPV(&Context, Info, R);
   TypeExprRewriter TER(&Context, Info, R);
   TypeArgumentAdder TPA(&Context, Info, R);
-  for (auto &D : TUD->decls()) {
-    V.TraverseDecl(D);
-    FV.TraverseDecl(D);
+  TranslationUnitDecl *TUD = Context.getTranslationUnitDecl();
+  for (const auto &D : TUD->decls()) {
     ECPV.TraverseDecl(D);
     if (AddCheckedRegions) {
       // Adding checked regions enabled!?
+      // TODO: Should checked region finding happen somewhere else? This is
```

Comment:
That's an interesting point. It used to be that Checked Region rewriting happened in one pass, but now it has been broken into two: Find regions, then rewrite them. The finding part still needs to come after constraint solving, so here's as good of a place as any?

Reply:
As long as only the CheckedRegionAdder needs to know if a block is a checked regions, it's fine to leave it here. We might want to have access to that information elsewhere eventualy for other analysis or diagnostic/benchmark output.